main/src/main/java/cgeo/geocaching/downloader/HillshadingTileDownloader.java

```diff
@@ -0,0 +1,118 @@
+        basicUpMatcher(uri, list, page, PATTERN_UP);
+
+        final MatcherWrapper matchDir = new MatcherWrapper(PATTERN_DIR, page);
+        while (matchDir.find()) {
+            final Download offlineMap = new Download(matchDir.group(1), Uri.parse(uri + matchDir.group(1)), true, "", "", offlineMapType, ICONRES_FOLDER);
+            list.add(offlineMap);
+        }
+
+        final MatcherWrapper matchMap = new MatcherWrapper(PATTERN_TILE, page);
+        while (matchMap.find()) {
+            final Download offlineMap = new Download(matchMap.group(1), Uri.parse(uri + matchMap.group(1)), false, matchMap.group(2), matchMap.group(3), offlineMapType, iconRes);
+            list.add(offlineMap);
+        }
+    }
+
+    @Nullable
+    @Override
+    protected Download checkUpdateFor(final @NonNull String page, final String remoteUrl, final String remoteFilename) {
+        final MatcherWrapper matchMap = new MatcherWrapper(PATTERN_TILE, page);
+        while (matchMap.find()) {
+            final String filename = matchMap.group(1);
+            if (filename.equals(remoteFilename)) {
+                return new Download(matchMap.group(2), Uri.parse(remoteUrl + "/" + filename), false, matchMap.group(3), matchMap.group(4), offlineMapType, iconRes);
+            }
+        }
+        return null;
+    }
+
+    // BRouter uses a single download page, need to map here to its fixed address
+    /*@Override
+    protected String getUpdatePageUrl(final String downloadPageUrl) {
+        return mapBase.toString();
+    }*/
+
+    @NonNull
+    public static HillshadingTileDownloader getInstance() {
+        return INSTANCE;
+    }
+
+    // used for area tile checking, see MapUtils
+    @WorkerThread
+    public HashMap<String, Download> getAvailableTiles() {
+        final HashMap<String, Download> tiles = new HashMap<>();
+
+        final String url = CgeoApplication.getInstance().getString(R.string.hillshading_downloadurl);
+        for (int i = 0; i <= 83; i++) {
+            tiles.putAll(getAvailableTilesSubfolder(url + "N" + String.format("%02d", i) + "/"));
+        }
+        for (int i = 1; i <= 90; i++) {
+            tiles.putAll(getAvailableTilesSubfolder(url + "S" + String.format("%02d", i) + "/"));
+        }
+        return tiles;
+    }
+
+    private HashMap<String, Download> getAvailableTilesSubfolder(final String url) {
+        final HashMap<String, Download> tiles = new HashMap<>();
+        final String page = Network.getResponseData(Network.getRequest(url));
+        final List<Download> list = new ArrayList<>();
+        if (page != null) {
+            analyzePage(Uri.parse(url), list, page);
+        }
+        for (Download download : list) {
+            tiles.put(download.getName(), download);
+        }
+        return tiles;
+    }
+
+    @Override
+    protected boolean verifiedBeforeCopying(final String filename, final Uri file) {
```

Comment:
This code is BRouter-specific. Probably we do not have a verification routine for downloaded hillshading tiles, so you can omit this method completely.